Gh 791 add named operation group label & black/whitelist

macenturalxl1 opened the pull request:
Hi - here is the UI feature to add a group name label to a NamedOperation and view it as a dynamically added prefix in the operation dropdown for Issue #791 

Let me know what you think of the design how to view grouped NamedOperations in the dropdown (I've attached a screenshot on the Issue what it looks like from this implementation). I wasn't sure how far to go with the design - so kept it simple and in a state that is easy to change, maybe at a later date as this new feature matures.

*I'll add the whitelist/blacklist in another Pull Request to make continous integration easier - as this feature is something ready to use by a User. Or if you prefer I can put it all in a bigger Pull Request.

macenturalxl1 commented:
Hi - just done some further investigation - in analytics-ui module at error.service.spec.ts:45:17, a new Error() is instanitated as part of the test that logs an error but does not fail the test nor brack the build.

It's the latest versionn of Sass (v4.13.1) that is breaking the build as it does not support the Travis environment

`Error: Node Sass does not yet support your current environment: Linux 64-bit with Unsupported runtime (83)`

https://github.com/sass/node-sass/releases/tag/v4.13.1

I'll look in to fixing it...

macenturalxl1 commented:
Figured it out:

The pipeline is using Node v14.0.0, and Sass v4.13.1 only supports Node up to v13.x.x

t11947 commented:
Hi @macenturalxl1, thanks for the update on your progress. How might things be affected if a hash were to be included in the name of a label (or named operation for that matter)?

Thanks for bringing that node issue to our attention, we'll get that fixed as a priority.

macenturalxl1 commented:
> Hi @macenturalxl1, thanks for the update on your progress. How might things be affected if a hash were to be included in the name of a label (or named operation for that matter)?

Hi, the hashtags is just UI visual presentation - the HTML views append a # to the beginning of the label to view in the DOM - in the controller/model the labels are stored exactly as the User inputs them - including when they are added to the AddNamedOperation request body.

You make a good point as it could be confusing for the User thinking they're added to the NamedOperation with a # suffix if they are concerned with how it is implemented and if they don't see the request being made from the network tab in debug - unless for the User is just a black box User and want to any some form of label tagging?

Let me know if this requires a redesign

macenturalxl1 commented:
Hi - any thoughts on the whiteList/blackList scenarios? > https://github.com/gchq/gaffer-tools/issues/791

I'll raise the white/blackList change in a separate PR.

macenturalxl1 commented:
Hi I've fixed the pipeline where the version of Node was causing a dependency in Analytics to break. 

I've also added the blacklist/whitelist feature to this using the rules confirmed in the Issue comments > https://github.com/gchq/gaffer-tools/issues/791

macenturalxl1 commented:
Hi - any feedback on this pull request?

t11947 commented on cd/verify.sh:
```diff
@@ -16,7 +16,7 @@ if [[ "$RELEASE" != 'true' ]] && [[ "$TRAVIS_PULL_REQUEST" != 'false' ]]; then
 
         # Update nodejs to the latest version
-        nvm install node
+        nvm install 13.13.0
```
It may be worth switching this to use the current LTS version as opposed to a specific version.

d47853 replied:
```suggestion
        nvm install --lts
```

t11947 replied:
After further consideration, it has been communicated that the lts version is incompatible with some items in the analytic ui project, as such we should perhaps keep the fixed version number for now. Alternatively, we could specify a conditional command specific to the analytics ui.

macenturalxl1 replied:
Hi - I tried with `lts` and Travis seems to terminate with no explanation why in the logs (I think it might hang and then timeout). I've also tried `--lts` and Travis doesn't recognise it as an option.

I've also tried reverting back to the original `nvm install node` that installs version 14.4.0 . But the 'latest' Sass version in Analytics is only compatible with versions 13.x.x and not 14.x.x - so I stuck with the latest 13.x.x version 

It's a temporary fix in Analytics to ensure the pipeline still builds for this Pull Request
